src/main/java/com/artipie/asto/s3/MultipartUpload.java

```diff
@@ -70,36 +83,33 @@ final class MultipartUpload {
      * @return Completion stage which is completed when responses received from S3 for all parts.
      */
     public CompletionStage<Void> upload(final Content content) {
-        final AtomicInteger part = new AtomicInteger();
-        return Flowable.fromPublisher(content).compose(
-            FlowableTransformers.bufferWhile(
-                new Predicate<ByteBuffer>() {
-                    private long sum;
-
-                    @Override
-                    public boolean test(final ByteBuffer buffer) {
-                        final int length = buffer.remaining();
-                        final boolean keep;
-                        if (this.sum + length > MultipartUpload.MIN_PART_SIZE) {
-                            this.sum = length;
-                            keep = false;
-                        } else {
-                            this.sum += length;
-                            keep = true;
+        final AtomicInteger counter = new AtomicInteger();
+        return Flowable.fromPublisher(content)
+            .concatMap(
+                buffer -> Flowable.fromPublisher(
+                    new Splitting(buffer, MultipartUpload.MIN_PART_SIZE).publisher()
+                )
+            ).map(
+                chunk -> {
+                    final int pnum = counter.incrementAndGet();
+                    return MultipartUpload.this.uploadPart(
+                        pnum,
+                        Flowable.just(chunk)
+                    ).<Void>thenApply(
```

Comment:
It would be better to use `thenAccept` here instead of  `thenApply` as `thenAccept` returns void itself